File: nakli-hub/internal/server/envelope.go
// Package server hosts the HTTP handlers, middleware, and the response
// envelope plumbing for nakli-hub. Wire shapes follow fabric-spec-001-v1.0.md.
package server

import (
	"context"
	"encoding/json"
	"net/http"
	"time"
)

// ProtocolVersion is the wire-protocol version string sent in every response.
const ProtocolVersion = "naklimesh/1.0"

// SuccessEnvelope is the JSON shape of a successful response.
type SuccessEnvelope struct {
	OK        bool        `json:"ok"`
	Data      interface{} `json:"data"`
	Freshness *Freshness  `json:"freshness,omitempty"`
}

// ErrorEnvelope is the JSON shape of an error response.
type ErrorEnvelope struct {
	OK    bool      `json:"ok"`
	Error ErrorBody `json:"error"`
}

// ErrorBody matches the protocol error shape.
type ErrorBody struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

// Freshness is the protocol's freshness object. For Phase 2a (no peers yet)
// we report zero peers and zero staleness.
type Freshness struct {
	AsOf         time.Time `json:"as_of"`
	PeersSynced  []string  `json:"peers_synced"`
	PeersMissing []string  `json:"peers_missing"`
	StalenessMs  int64     `json:"staleness_ms"`
}

// FreshnessNow returns a Freshness snapshot for the current moment with no peers.
func FreshnessNow(now time.Time) *Freshness {
	return &Freshness{
		AsOf:         now.UTC(),
		PeersSynced:  []string{},
		PeersMissing: []string{},
		StalenessMs:  0,
	}
}

// writeSuccess emits a SuccessEnvelope with status 200 (or status if provided).
func writeSuccess(w http.ResponseWriter, r *http.Request, status int, data interface{}, fresh *Freshness) []byte {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	body, _ := json.Marshal(SuccessEnvelope{OK: true, Data: data, Freshness: fresh})
	_, _ = w.Write(body)
	return body
}

// writeError emits an ErrorEnvelope with the appropriate HTTP status. Retryable
// is derived from the protocol's error catalogue (see fabric-spec §wire format).
func writeError(w http.ResponseWriter, r *http.Request, status int, code, message string, retryable bool) []byte {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	body, _ := json.Marshal(ErrorEnvelope{
		OK: false,
		Error: ErrorBody{
			Code:      code,
			Message:   message,
			Retryable: retryable,
		},
	})
	_, _ = w.Write(body)
	return body
}

// Common error codes (subset; full catalogue in fabric-spec §Wire format).
const (
	ErrGrantInvalid         = "grant_invalid"
	ErrGrantMissing         = "grant_missing"
	ErrGrantRevoked         = "grant_revoked"
	ErrScopeDenied          = "scope_denied"
	ErrCaveatUnmet          = "caveat_unmet"
	ErrIdempotencyConflict  = "idempotency_conflict"
	ErrNotFound             = "not_found"
	ErrConflict             = "conflict"
	ErrUnavailable          = "unavailable"
	ErrVersionMismatch      = "version_mismatch"
	ErrBadRequest           = "bad_request"
	ErrNotImplemented       = "not_implemented"
)

// Context keys used by middleware to pass per-request data to handlers.
type ctxKey int

const (
	ctxKeyRequestID ctxKey = iota
	ctxKeyGrantID
	ctxKeyPrincipal
	ctxKeyGrantBytes
	ctxKeyIdempotencyKey
)

// RequestID returns the per-request ULID assigned by middleware, or "" if
// the request is unauthenticated and the logging middleware did not run.
func RequestID(ctx context.Context) string { v, _ := ctx.Value(ctxKeyRequestID).(string); return v }

// GrantID returns the verified Grant's grant_id for the request.
func GrantID(ctx context.Context) string { v, _ := ctx.Value(ctxKeyGrantID).(string); return v }

// Principal returns the verified principal_id for the request.
func Principal(ctx context.Context) string { v, _ := ctx.Value(ctxKeyPrincipal).(string); return v }

// IdempotencyKey returns the request's X-Fabric-Idempotency-Key header value.
func IdempotencyKey(ctx context.Context) string { v, _ := ctx.Value(ctxKeyIdempotencyKey).(string); return v }

File: nakli-cf-worker/src/envelope.ts
// Wire-format response envelopes that mirror the Hub's
// (fabric-spec-001-v1.0.md §Wire format).

import { PROTOCOL_VERSION } from './env.js';

export interface Freshness {
  as_of: string;
  peers_synced: any[];
  peers_missing: any[];
  staleness_ms: number;
}

export function freshnessNow(): Freshness {
  return {
    as_of: new Date().toISOString(),
    peers_synced: [],
    peers_missing: [],
    staleness_ms: 0,
  };
}

const corsHeaders: Record<string, string> = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers':
    'Content-Type, X-Fabric-Grant, X-Fabric-Idempotency-Key, X-Fabric-Request-Id, ' +
    'X-Fabric-Version, X-Fabric-Agent-Id, X-Fabric-Device-Id, X-Fabric-Principal-Type, ' +
    'X-Fabric-Discharge',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Expose-Headers': 'X-Fabric-Version, X-Fabric-Request-Id',
};

function baseHeaders(extra: Record<string, string> = {}): Headers {
  const h = new Headers({
    'Content-Type': 'application/json',
    'X-Fabric-Version': PROTOCOL_VERSION,
    ...corsHeaders,
    ...extra,
  });
  return h;
}

export function corsResponse(): Response {
  return new Response(null, { status: 204, headers: baseHeaders() });
}

export function successResponse(
  data: unknown,
  freshness: Freshness | null = freshnessNow(),
  status = 200,
  extraHeaders: Record<string, string> = {},
): Response {
  const body = JSON.stringify({ ok: true, data, freshness });
  return new Response(body, { status, headers: baseHeaders(extraHeaders) });
}

export interface ErrorBody {
  code: ErrorCode;
  message: string;
  retryable: boolean;
}

export function errorResponse(
  code: ErrorCode,
  message: string,
  status: number,
  retryable = false,
  data?: unknown,
): Response {
  const env: Record<string, unknown> = {
    ok: false,
    error: { code, message, retryable },
  };
  if (data !== undefined) env.data = data;
  return new Response(JSON.stringify(env), { status, headers: baseHeaders() });
}

// Error codes catalogue — subset that the Worker emits.
export type ErrorCode =
  | 'grant_invalid'
  | 'grant_missing'
  | 'grant_revoked'
  | 'scope_denied'
  | 'caveat_unmet'
  | 'idempotency_conflict'
  | 'not_found'
  | 'conflict'
  | 'unavailable'
  | 'version_mismatch'
  | 'bad_request'
  | 'not_implemented'
  | 'rate_limited'
  | 'human_approval_required'
  | 'principal_retired';
